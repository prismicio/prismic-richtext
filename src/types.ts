export const enum NodeType {
	HEADING_1 = "heading1",
	HEADING_2 = "heading2",
	HEADING_3 = "heading3",
	HEADING_4 = "heading4",
	HEADING_5 = "heading5",
	HEADING_6 = "heading6",
	PARAGRAPH = "paragraph",
	PREFORMATTED = "preformatted",
	STRONG = "strong",
	EM = "em",
	LIST_ITEM = "list-item",
	ORDERED_LIST_ITEM = "o-list-item",
	LIST = "group-list-item",
	ORDERED_LIST = "group-o-list-item",
	IMAGE = "image",
	EMBED = "embed",
	HYPERLINK = "hyperlink",
	LABEL = "label",
	SPAN = "span",
}

export type RTNode = RTTextNode | RTImageNode | RTEmbedNode | RTListGroupNode;
export type RTSpanNode = RTLabelSpanNode | RTTextSpanNode | RTHyperlinkSpanNode;

interface RTSpanNodeBase {
	start: number;
	end: number;
}

interface RTLabelSpanNode extends RTSpanNodeBase {
	type: NodeType.LABEL;
	data: {
		label: string;
	};
}

interface RTTextSpanNode extends RTSpanNodeBase {
	type: NodeType.EM | NodeType.SPAN | NodeType.STRONG;
}

interface RTHyperlinkSpanNode extends RTSpanNodeBase {
	type: NodeType.HYPERLINK;
	data: RTHyperlinkData;
}

type RTHyperlinkData =
	| RTHyperlinkDocumentData
	| RTHyperlinkMediaData
	| RTHyperlinkWebData;

export const enum HyperlinkType {
	DOCUMENT = "Document",
	MEDIA = "Media",
	WEB = "Web",
}

interface RTHyperlinkDocumentData {
	link_type: HyperlinkType.DOCUMENT;
	id: string;
	type: string;
	tags: string[];
	slug: string;
	lang: string;
	uid: string | null;
}

type RTHyperlinkMediaData =
	| RTHyperlinkMediaDocumentData
	| RTHyperlinkMediaImageData;

export const enum HyperlinkMediaKind {
	IMAGE = "image",
	DOCUMENT = "document",
}

interface RTHyperlinkMediaDocumentData {
	link_type: HyperlinkType.MEDIA;
	kind: HyperlinkMediaKind.DOCUMENT;
	name: string;
	url: string;
	size: number;
}

interface RTHyperlinkMediaImageData {
	link_type: HyperlinkType.MEDIA;
	kind: HyperlinkMediaKind.IMAGE;
	name: string;
	url: string;
	size: number;
	height: number;
	width: number;
}

interface RTHyperlinkWebData {
	link_type: HyperlinkType.WEB;
	url: string;
	target: string | null;
}

export interface RTTextNode {
	type:
		| NodeType.HEADING_1
		| NodeType.HEADING_2
		| NodeType.HEADING_3
		| NodeType.HEADING_4
		| NodeType.HEADING_5
		| NodeType.HEADING_6
		| NodeType.PARAGRAPH
		| NodeType.EM
		| NodeType.LIST_ITEM
		| NodeType.ORDERED_LIST_ITEM
		| NodeType.PREFORMATTED
		| NodeType.SPAN
		| NodeType.STRONG;
	text: string;
	spans: RTSpanNode[];
}

interface RTImageNode {
	type: NodeType.IMAGE;
	url: string;
	alt: string;
	copyright: string;
	dimensions: {
		width: number;
		height: number;
	};
}

interface RTEmbedNode {
	type: NodeType.EMBED;
	oembed: Record<string, string | number | null> & {
		html: string;
	};
}

export interface RTListGroupNode {
	type: NodeType.LIST | NodeType.ORDERED_LIST;
	listItems: RTListGroupItemNode[];
}

export interface RTListGroupItemNode extends RTTextNode {
	type: NodeType.LIST_ITEM | NodeType.ORDERED_LIST_ITEM;
}

export interface Tree {
	key: string;
	children: TreeNode[];
}

export interface TreeNode {
	key: string;
	type: NodeType;
	text?: string;
	node: RTNode | RTSpanNode;
	children: TreeNode[];
}
