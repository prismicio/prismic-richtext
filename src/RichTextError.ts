export class RichTextError extends Error {}
