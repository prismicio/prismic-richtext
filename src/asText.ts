import { RTNode } from "./types";

export function asText(nodes: RTNode[], separator = " "): string {
	return nodes
		.map((node) => {
			if ("text" in node) {
				return node.text;
			}
		})
		.filter(Boolean)
		.join(separator);
}
