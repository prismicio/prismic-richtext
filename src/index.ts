export { asTree } from "./asTree";
export { asText } from "./asText";

export { serialize } from "./serialize";
export { wrapMapSerializer } from "./wrapMapSerializer";
export { composeSerializers } from "./composeSerializers";

export { RichTextNodeType as Element } from "@prismicio/types";

export type {
	RichTextFunctionSerializer,
	RichTextMapSerializer,
	RichTextMapSerializerFunction,
} from "./types";
