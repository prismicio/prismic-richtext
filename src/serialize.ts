import Tree from './tree';
import { Node, SpanNode, NodeElement } from './nodes';
import { RichTextBlock } from './richtext';

interface ParentNodeProps {
  [p: string]: any;
}

type Serializer<T> = (
  type: string,
  element: NodeElement,
  text: string | null,
  children: T[],
  index: number,
  parentNodeProps: ParentNodeProps
) => T;

function serializeNode<T>(
  parentNode: Node,
  serializer: Serializer<T>,
  index: number,
  htmlSerializer?: Serializer<T>,
  parentNodeProps?: ParentNodeProps
): T {
  function step(node: Node, idx: number): T {
    const text = node instanceof SpanNode ? node.text : null;
    const serializedChildren = node.children.reduce<T[]>((acc: T[], current: Node, i: number) => {
      return acc.concat([step(current, i)]);
    }, []);

    const maybeSerialized =
      htmlSerializer && htmlSerializer(node.type, node.element, text, serializedChildren, idx, parentNodeProps || {});
    return maybeSerialized || serializer(node.type, node.element, text, serializedChildren, idx, parentNodeProps || {});
  }

  return step(parentNode, index);
}

function fromRichText<T>(
  richText: RichTextBlock[],
  serialize: Serializer<T>,
  htmlSerializer?: Serializer<T>,
  parentNodeProps?: ParentNodeProps
): T[] {
  const tree = Tree.fromRichText(richText);
  return tree.children.map((node: Node, index: number) => {
    return serializeNode<T>(node, serialize, index, htmlSerializer, parentNodeProps);
  });
}

export default fromRichText;
