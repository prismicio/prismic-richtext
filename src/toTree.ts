import {
	NodeType,
	RTListGroupItemNode,
	RTNode,
	RTSpanNode,
	RTTextNode,
	Tree,
	TreeNode,
} from "./types";

export const uuid = (): string => {
	return (++uuid.i).toString();
};
uuid.i = 0;

export const toTree = (nodes: RTNode[]): Tree => {
	const preparedNodes = prepareNodes(nodes);

	const children: TreeNode[] = [];
	for (let i = 0; i < preparedNodes.length; i++) {
		children.push(nodeToTreeNode(preparedNodes[i]));
	}

	return {
		key: uuid(),
		children,
	};
};

const createTreeNode = (
	node: RTNode | RTSpanNode,
	children: TreeNode[] = [],
): TreeNode => {
	return {
		key: uuid(),
		type: node.type,
		text: "text" in node ? node.text : undefined,
		node,
		children,
	};
};

const createTextTreeNode = (text: string): TreeNode => {
	return createTreeNode({
		type: NodeType.span,
		text,
		spans: [],
	});
};

const prepareNodes = (nodes: RTNode[]): RTNode[] => {
	const mutNodes = nodes.slice(0);

	for (let i = 0; i < mutNodes.length; i++) {
		const node = mutNodes[i];

		if (node.type === NodeType.listItem || node.type === NodeType.oListItem) {
			const items: RTListGroupItemNode[] = [node as RTListGroupItemNode];

			while (mutNodes[i + 1] && mutNodes[i + 1].type === node.type) {
				items.push(mutNodes[i + 1] as RTListGroupItemNode);
				mutNodes.splice(i, 1);
			}

			mutNodes[i] = {
				type: node.type === NodeType.listItem ? NodeType.list : NodeType.oList,
				items,
			};
		}
	}

	return mutNodes;
};

const nodeToTreeNode = (node: RTNode): TreeNode => {
	if ("text" in node) {
		return createTreeNode(
			node,
			textNodeSpansToTreeNodeChildren(node.spans, node),
		);
	}

	if ("items" in node) {
		const children: TreeNode[] = [];
		for (let i = 0; i < node.items.length; i++) {
			children.push(nodeToTreeNode(node.items[i]));
		}

		return createTreeNode(node, children);
	}

	return createTreeNode(node);
};

const textNodeSpansToTreeNodeChildren = (
	spans: RTSpanNode[],
	node: RTTextNode,
	parentSpan?: RTSpanNode,
): TreeNode[] => {
	if (!spans.length) {
		return [createTextTreeNode(node.text)];
	}

	const children: TreeNode[] = [];

	for (let i = 0; i < spans.length; i++) {
		const span = spans[i];
		const parentSpanStart = (parentSpan && parentSpan.start) || 0;
		const spanStart = span.start - parentSpanStart;
		const spanEnd = span.end - parentSpanStart;

		const childSpans: RTSpanNode[] = [];
		for (let j = 0; j < spans.length; j++) {
			const siblingSpan = spans[j];

			if (
				siblingSpan !== span &&
				siblingSpan.start >= span.start &&
				siblingSpan.end <= span.end
			) {
				childSpans.push(siblingSpan);
				spans.splice(j, 1);
				j--;
			}
		}

		if (i === 0 && spanStart > 0) {
			children.push(createTextTreeNode(node.text.slice(0, spanStart)));
		}

		children.push(
			createTreeNode(
				span,
				textNodeSpansToTreeNodeChildren(
					childSpans,
					{
						...node,
						text: node.text.slice(spanStart, spanEnd),
					},
					span,
				),
			),
		);

		if (spanEnd < node.text.length) {
			children.push(
				createTextTreeNode(
					node.text.slice(
						spanEnd,
						spans[i + 1] ? spans[i + 1].start - parentSpanStart : undefined,
					),
				),
			);
		}
	}

	return children;
};
