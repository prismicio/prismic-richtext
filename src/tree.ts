import flatten from "ramda/es/flatten";
import init from "ramda/es/init";
import last from "ramda/es/last";
import sortWith from "ramda/es/sortwith";
import sortBy from "ramda/es/sortby";
import uuid from './uuid';
import { RichTextSpan, RichTextBlock } from './richtext';
import { NODE_TYPES, PRIORITIES } from './types';
import {
  Boundaries,
  Node,
  SpanNode,
  TextNode,
  BlockNode,
  ListBlockNode,
  ListItemBlockNode,
  OrderedListBlockNode,
  OrderedListItemBlockNode,
} from './nodes';

export interface Group {
  elected: SpanNode;
  others: SpanNode[];
}

interface SlicedNode {
  inner: SpanNode;
  outer?: SpanNode;
}

interface PartitionedGroup {
  inner: SpanNode[];
  outer: SpanNode[];
}

function sortByPriorities(nodes: SpanNode[]): SpanNode[] {
  return nodes.sort((nodeA: SpanNode, nodeB: SpanNode): number => {
    if (nodeA.isParentOf(nodeB)) {
      return -1;
    } else if (nodeB.isParentOf(nodeA)) {
      return 1;
    } else {
      const result = PRIORITIES[nodeA.type] - PRIORITIES[nodeB.type];
      return (result === 0) ? (nodeA.text.length - nodeB.text.length) : result;
    }
  });
}

function sliceNode(text: string, elected: SpanNode, node: SpanNode): SlicedNode {
  if (node.start < elected.start) {
    return {
      inner: SpanNode.slice(node, elected.start, node.end, text),
      outer: SpanNode.slice(node, node.start, elected.start, text),
    };
  } else if (node.end > elected.end) {
      return {
        inner: SpanNode.slice(node, node.start, elected.end, text),
        outer: SpanNode.slice(node, elected.end, node.end, text),
      };
  } else {
    return {
      inner: node,
    };
  }
}

function partitionGroup(text: string, group: Group): SpanNode[] {
  const partitioned = group.others.reduce<PartitionedGroup>(({ inner: innerAcc, outer: outerAcc }, node) => {
    const slicedNode = sliceNode(text, group.elected, node);
    return {
      inner: innerAcc.concat(slicedNode.inner),
      outer: slicedNode.outer ? outerAcc.concat(slicedNode.outer) : outerAcc,
    };
  }, { inner: [], outer: [] });

  const { inner, outer } = partitioned;
  const head = group.elected.setChildren(buildTreeAndFill(text, inner, group.elected.boundaries()));
  return [head].concat(buildTree(text, outer));
}

function groupWith(p: (nodeA: SpanNode, nodeB: SpanNode) => boolean, nodes: SpanNode[]): SpanNode[][] {
  return nodes.reduce<SpanNode[][]>((groups: SpanNode[][], node: SpanNode) => {
    const previousGroup = last(groups);
    if (previousGroup) {
      const included = previousGroup.some(nodeGroup => nodeGroup.isParentOf(node));
      if (included) {
        return init(groups).concat([previousGroup.concat(node)]);
      } else {
        const previousNode = last(previousGroup);
        if (previousNode && p(previousNode, node)) {
          return init(groups).concat([previousGroup.concat(node)]);
        } else {
          return groups.concat([[node]]);
        }
      }
    } else {
      return [[node]];
    }
  }, []);
}

function groupNodes(nodes: SpanNode[]): SpanNode[][] {
  const sortByStart = (nodeA: SpanNode, nodeB: SpanNode) => nodeA.start - nodeB.start;
  const sortByEnd = (nodeA: SpanNode, nodeB: SpanNode) => nodeA.end - nodeB.end;
  const sortedNodes = sortWith([sortByStart, sortByEnd], nodes);
  return groupWith((nodeA: SpanNode, nodeB: SpanNode) => nodeA.end >= nodeB.start, sortedNodes);
}

function electNode(candidates: SpanNode[]): Group {
  if (candidates.length === 0) {
    throw new Error('Unable to elect node on empty list');
  } else {
    const [elected, ...others] = sortByPriorities(candidates);
    return { elected, others };
  }
}

function fill(text: string, nodes: SpanNode[], boundaries: Boundaries): SpanNode[] {
  return nodes.reduce((acc: SpanNode[], node, index) => {
    let result: SpanNode[] = [];
    const fillStart = index === 0 && node.start > boundaries.lower;
    const fillEnd = index === nodes.length - 1 && boundaries.upper > node.end;

    if (fillStart) {
      const textNode = new TextNode(boundaries.lower, node.start, text.slice(boundaries.lower, node.start));
      result = result.concat(textNode);
    } else {
      const previousNode = nodes[index - 1];
      if (previousNode) {
        if (node.start > previousNode.end) {
          const subtext = text.slice(previousNode.end, node.start);
          const textNode = new TextNode(previousNode.end, node.start, subtext);
          result = result.concat(textNode);
        }
      }
    }

    result = result.concat(node);

    if (fillEnd) {
      const textNode = new TextNode(node.end, boundaries.upper, text.slice(node.end, boundaries.upper));
      result = result.concat(textNode);
    }

    return acc.concat(result);
  }, []);
}

function buildTreeAndFill(text: string, nodes: SpanNode[], boundaries: Boundaries): SpanNode[] {
  if (nodes.length > 0) {
    const tree = buildTree(text, nodes);
    return fill(text, tree, boundaries);
  } else {
    const subtext = text.slice(boundaries.lower, boundaries.upper);
    return [new TextNode(boundaries.lower, boundaries.upper, subtext)];
  }
}

function buildTree(text: string, nodes: SpanNode[]): SpanNode[] {
  const sortedNodes: SpanNode[] = sortBy((node: SpanNode) => node.start, nodes);
  const groups: SpanNode[][] = groupNodes(sortedNodes);
  const postElection: Group[] = groups.map(electNode);
  const tree: SpanNode[] = flatten<SpanNode>(postElection.map(group => partitionGroup(text, group)));
  return sortBy((node: SpanNode) => node.start, tree);
}

function processTextBlock(block: RichTextBlock): SpanNode[] {
  const nodes = block.spans.map((span) => {
    const text = block.text.slice(span.start, span.end);
    return new SpanNode(span.start, span.end, span.type, text, [], span);
  });
  const boundaries = { lower: 0, upper: block.text.length };
  return buildTreeAndFill(block.text, nodes, boundaries);
}

export default class Tree {
  key: string;
  children: Node[];

  static NODE_TYPES = NODE_TYPES;

  static fromRichText(richText: RichTextBlock[]): Tree {
    return {
      key: uuid(),
      children: richText.reduce<Node[]>((acc, block, index) => {
        if (RichTextBlock.isEmbedBlock(block.type) || RichTextBlock.isImageBlock(block.type)) {
          return acc.concat(new BlockNode(block.type, block));
        } else {
          const textNodes = processTextBlock(block);
          const previousBlock = acc[acc.length - 1];
          if (RichTextBlock.isListItem(block.type) && previousBlock && previousBlock instanceof ListBlockNode) {
            const listItem = new ListItemBlockNode(block, textNodes);
            const updatedPreviousBlock = previousBlock.addChild(listItem);
            return init(acc).concat(updatedPreviousBlock);
          } else if (RichTextBlock.isOrderedListItem(block.type) && previousBlock && previousBlock instanceof OrderedListBlockNode) {
            const orderedListItem = new OrderedListItemBlockNode(block, textNodes);
            const updatedPreviousBlock = previousBlock.addChild(orderedListItem);
            return init(acc).concat(updatedPreviousBlock);
          } else if (RichTextBlock.isListItem(block.type)) {
            const listItem = new ListItemBlockNode(block, textNodes);
            const list = new ListBlockNode(RichTextBlock.emptyList(), [listItem]);
            return acc.concat(list);
          } else if (RichTextBlock.isOrderedListItem(block.type)) {
            const orderedListItem = new OrderedListItemBlockNode(block, textNodes);
            const orderedList = new OrderedListBlockNode(RichTextBlock.emptyOrderedList(), [orderedListItem]);
            return acc.concat(orderedList);
          } else {
            return acc.concat(new BlockNode(block.type, block, textNodes));
          }
        }
      }, []),
    };
  }
}
