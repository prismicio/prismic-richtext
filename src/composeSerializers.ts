import { RichTextError } from "./RichTextError";
import { RichTextFunctionSerializer } from "./types";

export const composeSerializers = <T>(
	...serializers: [
		RichTextFunctionSerializer<T>,
		...RichTextFunctionSerializer<T>[]
	]
) => {
	return (...args: Parameters<RichTextFunctionSerializer<T>>): T => {
		for (let i = 0; i < serializers.length; i++) {
			const res = serializers[i](...args);

			if (res !== null) {
				return res;
			}
		}

		throw new RichTextError(
			`Rich Text Node of type ${args[0]} does not have a serializer`,
		);
	};
};
