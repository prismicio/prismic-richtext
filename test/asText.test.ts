import { it, expect } from "vitest";

import { createRichTextFixtures } from "./__testutils__/createRichTextFixtures";

import { asText } from "../src";

it("returns a string representation of a rich text field value", () => {
	const richTextFixtures = createRichTextFixtures();

	expect(asText(richTextFixtures.en)).toMatchSnapshot();
});

it("allows for custom separator", () => {
	const richTextFixtures = createRichTextFixtures();

	expect(asText(richTextFixtures.en, "SEPARATOR")).toMatchSnapshot();
});
