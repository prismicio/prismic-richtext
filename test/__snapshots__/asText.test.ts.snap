// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`allows for custom separator 1`] = `
"This block has a line
break just before the word \\"break\\"SEPARATORText before with a multi-block span!SEPARATORHeading 1 bold text italic still bold normalSEPARATORParagraphSEPARATORA paragraph with a nested labelSEPARATORFirst ordered list itemSEPARATORSecond ordered list itemSEPARATORThird ordered list itemSEPARATORBold and now italic but not anymore value in the last list itemSEPARATORParagraphSEPARATORFirst list itemSEPARATORSecond list itemSEPARATORThird list itemSEPARATORParagraphSEPARATORA web linkSEPARATORA document linkSEPARATORA media linkSEPARATORA non-image media linkSEPARATORA paragraph with a nested document linkSEPARATORParagraphSEPARATORParagraph"
`;

exports[`returns a string representation of a rich text field value 1`] = `
"This block has a line
break just before the word \\"break\\" Text before with a multi-block span! Heading 1 bold text italic still bold normal Paragraph A paragraph with a nested label First ordered list item Second ordered list item Third ordered list item Bold and now italic but not anymore value in the last list item Paragraph First list item Second list item Third list item Paragraph A web link A document link A media link A non-image media link A paragraph with a nested document link Paragraph Paragraph"
`;
